dali/pipeline/executor/workspace_policy.h

```diff
@@ -250,15 +252,188 @@ struct JIT_WS_Policy {
+                                    mixed_op_stream, gpu_op_stream, mixed_op_events);
+      }
+    }
+
+    for (int support_id = 0; support_id < depths_[OpType::SUPPORT]; support_id++) {
+      for (int cpu_id = 0; cpu_id < depths_[OpType::CPU]; cpu_id++) {
+        for (int mixed_id = 0; mixed_id < depths_[OpType::MIXED]; mixed_id++) {
+          // Get sequential index and prepare space all MIXED Ops
+          auto queue_idxs = QueueIdxs{support_id, cpu_id, mixed_id, 0};
+          PlaceWorkspace<OpType::MIXED>(mixed_workspaces_, queue_idxs, graph, tensor_to_store_queue,
+                                        mixed_op_stream, gpu_op_stream, mixed_op_events);
+        }
+      }
+    }
+
+    // we reuse the loop for Mixed with GPU as they are in sync
+    for (int support_id = 0; support_id < depths_[OpType::SUPPORT]; support_id++) {
+      for (int cpu_id = 0; cpu_id < depths_[OpType::CPU]; cpu_id++) {
+        for (int mixed_id = 0; mixed_id < depths_[OpType::MIXED]; mixed_id++) {
+          // Get sequential index and prepare space all GPU Ops
+          auto queue_idxs = QueueIdxs{support_id, cpu_id, mixed_id, mixed_id};
+          PlaceWorkspace<OpType::GPU, OpType::MIXED>(gpu_workspaces_, queue_idxs, graph,
+                                                     tensor_to_store_queue, mixed_op_stream,
+                                                     gpu_op_stream, mixed_op_events);
+        }
+      }
+    }
+  }
+
+  template <OpType op_type>
+  ws_t<op_type> GetWorkspace(QueueIdxs idxs, const OpGraph &graph, OpPartitionId partition_idx) {
+    // Handle MIXED and GPU indexing together
+    auto index_for_op_type = op_type == OpType::GPU ? OpType::MIXED : op_type;
+    int sequential_ws_idx = SequentialIndex(idxs, depths_, index_for_op_type);
+    auto &workspaces = GetWorkspacesCollection<op_type>();
+    return workspaces[sequential_ws_idx][partition_idx];
+  }
+
+  template <OpType op_type>
+  ws_t<op_type> GetWorkspace(QueueIdxs idxs, const OpGraph &graph, const OpNode &node) {
+    DALI_ENFORCE(node.op_type == op_type,
+                 "Wrong variant of method selected. OpType does not match.");
+    return GetWorkspace<op_type>(idxs, graph, node.partition_index);
+  }
+
+ private:
+  StageQueues depths_;
+  // ws_id -> op_id -> workspace
+  std::vector<std::vector<SupportWorkspace>> support_workspaces_;
+  std::vector<std::vector<HostWorkspace>> cpu_workspaces_;
+  std::vector<std::vector<MixedWorkspace>> mixed_workspaces_;
+  std::vector<std::vector<DeviceWorkspace>> gpu_workspaces_;
+
+  template <OpType op_type>
+  using ws_collection_t = typename std::vector<std::vector<op_type_to_workspace_t<op_type>>>;
+
+  // Access one of `support_workspaces_`, `cpu_workspaces_`, `mixed_workspaces_`, `gpu_workspaces_`
+  // based on op_type. Below are specialization for the sake of simplicity
+  template <OpType op_type>
+  ws_collection_t<op_type> &GetWorkspacesCollection() {
+    return detail::get<ws_collection_t<op_type>&>(
```

Comment:
Why this, when you have full specialized templates for all the possible template arguments?